File: examples/gripper/gripper.cc
/**
 * Copyright 2022-2025 lebai.ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <iostream>
#include <cstring>
#include <memory>
#include <stdexcept>
#include <string>
#include "lebai/gripper.hh"

int main(int argc, char* argv[]) {
	// Check command line arguments
	if (argc < 2 || argc > 3) {
		std::cerr << "Usage: " << argv[0] << " <serial_port> [position]" << std::endl;
		std::cerr << "Example (Windows): " << argv[0] << " COM3 50" << std::endl;
		std::cerr << "Example (Linux): " << argv[0] << " /dev/ttyUSB0 50" << std::endl;
		return 1;
	}

	std::string port_name = argv[1];
	unsigned int position = 50;
	if (argc == 3) {
		try {
			size_t parsed = 0;
			unsigned long value = std::stoul(argv[2], &parsed, 10);
			if (argv[2][0] == '-' || parsed != std::strlen(argv[2]) ||
			    value > 100) {
				throw std::invalid_argument("trailing characters");
			}
			position = static_cast<unsigned int>(value);
		}
		catch (const std::exception&) {
			std::cerr << "Invalid position: " << argv[2] << std::endl;
			return 1;
		}
	}

	try {
		std::cout << "Connecting to gripper on port: " << port_name << std::endl;

		// Create gripper instance
		// This will establish Modbus RTU communication over RS485
		// with 115200 baud rate, 8N1, Modbus address 1
		auto gripper = std::make_unique<lebai::l_master::Gripper>(port_name);

		std::cout << "Gripper connected successfully!" << std::endl;
		std::cout << "Setting gripper position to: " << position << std::endl;
		gripper->set_position(position);
	}
	catch (const std::exception& e) {
		std::cerr << "Error: " << e.what() << std::endl;
		return 1;
	}
	catch (...) {
		std::cerr << "Unknown error occurred" << std::endl;
		return 1;
	}

	return 0;
}

File: examples/gripper/gripper.py
# Copyright 2022-2025 lebai.ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import sys

def main():
    # Check command line arguments
    if len(sys.argv) < 2 or len(sys.argv) > 3:
        print(f"Usage: {sys.argv[0]} <serial_port> [position]", file=sys.stderr)
        print(f"Example (Windows): {sys.argv[0]} COM3 50", file=sys.stderr)
        print(f"Example (Linux): {sys.argv[0]} /dev/ttyUSB0 50", file=sys.stderr)
        return 1

    port_name = sys.argv[1]
    position = 50
    if len(sys.argv) == 3:
        try:
            if sys.argv[2].startswith("-"):
                raise ValueError()
            position = int(sys.argv[2], 10)
            if position > 100:
                raise ValueError()
        except ValueError:
            print(f"Invalid position: {sys.argv[2]}", file=sys.stderr)
            return 1

    try:
        from pylebai import gripper

        print(f"Connecting to gripper on port: {port_name}")

        # Create gripper instance
        # This will establish Modbus RTU communication over RS485
        # with 115200 baud rate, 8N1, Modbus address 1
        direct_gripper = gripper.Gripper(port_name)

        print("Gripper connected successfully!")
        print(f"Setting gripper position to: {position}")
        direct_gripper.set_position(position)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        return 1

    return 0

if __name__ == "__main__":
    sys.exit(main())
